src/Storages/MaterializedView/tests/gtest_timers.cpp

```diff
@@ -0,0 +1,27 @@
+#include <gtest/gtest.h>
+
+#include <Storages/MaterializedView/RefreshTimers.h>
+#include <Parsers/ASTTimeInterval.h>
+
+using namespace DB;
+
+TEST(Timers, AfterTimer)
+{
+    using namespace std::chrono;
+
+    auto interval = std::make_shared<ASTTimeInterval>();
+    interval->kinds = {
+        {IntervalKind::Week, 2},
+        {IntervalKind::Day, 3},
+        {IntervalKind::Minute, 15},
+    };
+    RefreshAfterTimer timer(interval.get());
+
+    sys_days date_in = 2023y / January / 18d;
+    auto secs_in = date_in + 23h + 57min;
+
+    sys_days date_out = 2023y / February / 5d;
+    auto secs_out = date_out + 0h + 12min;
+
+    ASSERT_EQ(secs_out, timer.after(secs_in));
+}
```

Comment:
Consider adding a simple test that generates lots (1M?) of random parameters and timestamps and checks that `timer.after(timer.after(t) + 1ns) >= timer.after(t) + period / 2`, i.e. that we won't schedule multiple refreshes in quick succession. (Because it's not obvious to me that there isn't some corner case where things would get rounded different ways or something.)